src/storage/concurrency_manager/mod.rs

```diff
@@ -0,0 +1,155 @@
+//! to ensure serializability. Lock information can be also stored in the
+//! manager and reading requests can check if these locks block the read.
+
+mod key_handle;
+mod lock_table;
+
+pub use self::key_handle::{KeyHandle, KeyHandleGuard};
+pub use self::lock_table::LockTable;
+
+use std::{
+    mem::{self, MaybeUninit},
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+use txn_types::{Key, Lock, TimeStamp};
+
+// TODO: Currently we are using a Mutex<BTreeMap> to implement the handle table.
+// In the future we should replace it with a concurrent ordered map.
+// Pay attention that the async functions of ConcurrencyManager should not hold
+// the mutex.
+#[derive(Clone)]
+pub struct ConcurrencyManager {
+    max_read_ts: Arc<AtomicU64>,
+    lock_table: LockTable,
+}
+
+impl ConcurrencyManager {
+    pub fn new(latest_ts: TimeStamp) -> Self {
+        ConcurrencyManager {
+            max_read_ts: Arc::new(AtomicU64::new(latest_ts.into_inner())),
+            lock_table: LockTable::default(),
+        }
+    }
+
+    pub fn max_read_ts(&self) -> TimeStamp {
+        TimeStamp::new(self.max_read_ts.load(Ordering::SeqCst))
+    }
+
+    /// Acquires a mutex of the key and returns an RAII guard. When the guard goes
+    /// out of scope, the mutex will be unlocked.
+    ///
+    /// The guard can be used to store Lock in the table. The stored lock
+    /// is visible to `read_key_check` and `read_range_check`.
+    pub async fn lock_key(&self, key: &Key) -> KeyHandleGuard {
+        self.lock_table.lock_key(key).await
+    }
+
+    /// Acquires mutexes of the keys and returns the RAII guards. The order of the
+    /// guards is the same with the given keys.
+    ///
+    /// The guards can be used to store Lock in the table. The stored lock
+    /// is visible to `read_key_check` and `read_range_check`.
+    pub async fn lock_keys(&self, keys: impl Iterator<Item = &Key>) -> Vec<KeyHandleGuard> {
+        let mut keys_with_index: Vec<_> = keys.enumerate().collect();
+        // To prevent deadlock, we sort the keys and lock them one by one.
+        keys_with_index.sort_by_key(|(_, key)| *key);
+        let mut result: Vec<MaybeUninit<KeyHandleGuard>> = Vec::new();
+        result.resize_with(keys_with_index.len(), || MaybeUninit::uninit());
+        for (index, key) in keys_with_index {
+            result[index] = MaybeUninit::new(self.lock_table.lock_key(key).await);
+        }
+        #[allow(clippy::unsound_collection_transmute)]
+        unsafe {
+            mem::transmute(result)
+        }
+    }
+
+    /// Checks if there is a memory lock of the key which blocks the read.
+    /// The given `check_fn` should return false iff the lock passed in
+    /// blocks the read.
+    ///
+    /// It will also updates the max_read_ts.
+    pub fn read_key_check<E>(
+        &self,
+        key: &Key,
+        ts: TimeStamp,
+        check_fn: impl FnOnce(&Lock) -> Result<(), E>,
```

Comment:
Why do we need to take check_fn? Rather than just considering a key locked if the lock exists?

Reply:
The logic of checking locks is not that simple. We need to compare the timestamps, and also, when the read request provides a `TsSet`, we should ignore the locks whose ts is in that set. (That's an optimization for large txns)